snapshots/windows/cimfs.go

```diff
@@ -0,0 +1,276 @@
+	return s.legacySn.Stat(ctx, key)
+}
+
+func (s *cimfsSnapshotter) Update(ctx context.Context, info snapshots.Info, fieldpaths ...string) (snapshots.Info, error) {
+	return s.legacySn.Update(ctx, info, fieldpaths...)
+}
+
+func (s *cimfsSnapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
+	legacyUsage, err := s.legacySn.Usage(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	defer t.Rollback()
+
+	id, _, _, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, fmt.Errorf("get snapshot info: %w", err)
+	}
+
+	if ok, err := s.isCimLayer(ctx, key); err != nil {
+		return snapshots.Usage{}, err
+	} else if ok {
+		cimUsage, err := cimfs.GetCimUsage(ctx, getCimLayerPath(s.cimDir, id))
+		if err != nil {
+			return snapshots.Usage{}, err
+		}
+		legacyUsage.Size += int64(cimUsage)
+	}
+	return legacyUsage, nil
+}
+
+func (s *cimfsSnapshotter) Prepare(ctx context.Context, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	m, err := s.legacySn.createSnapshot(ctx, snapshots.KindActive, key, parent, opts)
+	if err != nil {
+		return m, err
+	}
+	return s.createCimfsMounts(ctx, m, key, opts...)
+}
+
+func (s *cimfsSnapshotter) View(ctx context.Context, key, parent string, opts ...snapshots.Opt) ([]mount.Mount, error) {
+	m, err := s.legacySn.createSnapshot(ctx, snapshots.KindView, key, parent, opts)
+	if err != nil {
+		return m, err
+	}
+	return s.createCimfsMounts(ctx, m, key, opts...)
+}
+
+func (s *cimfsSnapshotter) createCimfsMounts(ctx context.Context, m []mount.Mount, key string, opts ...snapshots.Opt) (_ []mount.Mount, retErr error) {
+	if len(m) != 1 {
+		return m, fmt.Errorf("expected exactly 1 mount from legacy windows snapshotter, found %d", len(m))
+	}
+	m[0].Type = "cimfs"
+	return m, nil
+}
+
+// Mounts returns the cimfs mounts for the snapshot identified by key.
+//
+// This can be used to recover mounts after calling View or Prepare.
+func (s *cimfsSnapshotter) Mounts(ctx context.Context, key string) ([]mount.Mount, error) {
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return nil, err
+	}
+	defer t.Rollback()
+
+	sn, err := storage.GetSnapshot(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("get snapshot: %w", err)
+	}
+
+	mounts, err := s.legacySn.mounts(sn, key), nil
+	if err != nil {
+		return nil, err
+	}
+	return s.createCimfsMounts(ctx, mounts, key)
+}
+
+func (s *cimfsSnapshotter) Commit(ctx context.Context, name, key string, opts ...snapshots.Opt) error {
+	ctx, t, err := s.legacySn.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
```

Comment:
Is this a warning or an error? If the rollback transaction fails is this an error?